packages/cli/src/util/dev/server.ts

```diff
@@ -693,11 +694,42 @@ export default class DevServer {
     buildEnv['NOW_REGION'] = 'dev1';
     allEnv['NOW_REGION'] = 'dev1';
 
-    // mirror how VERCEL_REGION is injected in prod/preview
-    // only inject in `runEnvs`, because `allEnvs` is exposed to dev command
-    // and should not contain VERCEL_REGION
     if (this.projectSettings?.autoExposeSystemEnvs) {
+      // mirror how VERCEL_REGION is injected in prod/preview
+      // only inject in `runEnvs`, because `allEnvs` is exposed to dev command
+      // and should not contain VERCEL_REGION
       runEnv['VERCEL_REGION'] = 'dev1';
+
+      // expose VERCEL_ENV and VERCEL_URL to dev command
+      allEnv['VERCEL_ENV'] = 'development';
+      if (this.address.host) {
+        allEnv['VERCEL_URL'] = this.address.host;
+      }
+
+      // automaticaly expose VERCEL_URL, VERCEL_ENV
+      // as prefixed (framework) env vars
+      if (this.projectSettings?.framework) {
+        const frameworkSlug = this.projectSettings.framework;
+        const framework = frameworkList.find(f => f.slug === frameworkSlug);
+        const envPrefix = framework?.envPrefix;
+
+        if (envPrefix) {
+          [buildEnv, runEnv, allEnv] = await Promise.all([
+            {
+              ...buildEnv,
+              ...getPrefixedEnvVars({ envPrefix: envPrefix, envs: buildEnv }),
+            },
+            {
+              ...runEnv,
+              ...getPrefixedEnvVars({ envPrefix: envPrefix, envs: runEnv }),
+            },
+            {
+              ...allEnv,
+              ...getPrefixedEnvVars({ envPrefix: envPrefix, envs: allEnv }),
```

Comment:
I'm pretty sure you only need to inject this in `buildEnv` since the prefixed env vars are only provided at build time

Reply:
@styfle setting these system env vars to `buildEnv` and doing this only for `buildEnv` didn't make them available to `vc dev` unless I assign them to `allEnv` this way (referencing them in user code always output `undefined`).

Reply:
You'll need to assign to `buildEnv` prior to `allEnv` being created

Reply:
In fact, that looks like a bug with the existing `VERCEL_REGION` too